core/drivers/xiphera_trng.c

```diff
@@ -37,30 +37,31 @@ static unsigned int trng_lock = SPINLOCK_UNLOCK;
 
 static vaddr_t xiphera_trng_base;
 
-static uint32_t xiphera_trng_read32(void)
+static bool xiphera_trng_random_available(void)
 {
-	uint32_t value = 0;
-	uint32_t exceptions = 0;
 	uint32_t status = 0;
 
-	while (true) {
-		/* Wait until we have value available */
-		status = io_read32(xiphera_trng_base + STATUS_REG);
-		if (status != TRNG_NEW_RAND_AVAILABLE)
-			continue;
+	status = io_read32(xiphera_trng_base + STATUS_REG);
 
-		value = io_read32(xiphera_trng_base + RAND_REG);
+	return status == TRNG_NEW_RAND_AVAILABLE;
+}
 
-		/*
-		 * Ack that RNG value has been consumed and trigger new one to
-		 * be generated
-		 */
-		io_write32(xiphera_trng_base + CONTROL_REG, HOST_TO_TRNG_READ);
-		io_write32(xiphera_trng_base + CONTROL_REG,
-			   HOST_TO_TRNG_ENABLE);
+static uint32_t xiphera_trng_read32(void)
+{
+	uint32_t value = 0;
 
-		break;
-	}
+	/* Sanity check that we actually have valid RNG available */
+	if (!xiphera_trng_random_available())
+		panic("RNG access error");
```

Comment:
IIUC, it is possible for `xiphera_trng_random_available()` to return false more or less at anytime (given the read loop that restores interrupts below). So how is it guaranteed that this particular call cannot fail?

Reply:
Looked at the VHDL for CPU interface part (other part is protected). There is a state machine in here. I don't think this can never fail. Will take it out so there is no sanity check for it.

Reply:
Removed and updated the PR.